demos/ota/ota_demo_core_http/ota_demo_core_http.c

```diff
@@ -0,0 +1,1620 @@
+ */
+
+#define OTA_NETWORK_BUFFER_SIZE          ( otaconfigFILE_BLOCK_SIZE + OTA_MAX_URL_SIZE + 128 )
+
+/**
+ * @brief The maximum number of retries for connecting to server.
+ */
+#define CONNECTION_RETRY_MAX_ATTEMPTS    ( 5U )
+
+/* Check that size of the user buffer is defined. */
+#ifndef USER_BUFFER_LENGTH
+    #define USER_BUFFER_LENGTH    ( 2048 )
+#endif
+
+/**
+ * @brief The MQTT metrics string expected by AWS IoT.
+ */
+#define METRICS_STRING           "?SDK=" OS_NAME "&Version=" OS_VERSION "&Platform=" HARDWARE_PLATFORM_NAME "&OTALib=" OTA_LIB
+
+/**
+ * @brief The length of the MQTT metrics string expected by AWS IoT.
+ */
+#define METRICS_STRING_LENGTH    ( ( uint16_t ) ( sizeof( METRICS_STRING ) - 1 ) )
+
+
+#ifdef CLIENT_USERNAME
+
+/**
+ * @brief Append the username with the metrics string if #CLIENT_USERNAME is defined.
+ *
+ * This is to support both metrics reporting and username/password based client
+ * authentication by AWS IoT.
+ */
+    #define CLIENT_USERNAME_WITH_METRICS    CLIENT_USERNAME METRICS_STRING
+#endif
+
+/*-----------------------------------------------------------*/
+
+/* Linkage for error reporting. */
+extern int errno;
+
+/**
+ * @brief Struct for firmware version.
+ */
+const AppVersion32_t appFirmwareVersion =
+{
+    .u.x.major = APP_VERSION_MAJOR,
+    .u.x.minor = APP_VERSION_MINOR,
+    .u.x.build = APP_VERSION_BUILD,
+};
+
+/* Each compilation unit must define the NetworkContext struct. */
+struct NetworkContext
+{
+    OpensslParams_t * pParams;
+};
+
+/**
+ * @brief Network connection context used in this demo.
+ */
+static NetworkContext_t networkContext;
+
+/**
+ * @brief Network connection context used for HTTP connection.
+ */
+static NetworkContext_t networkContextHttp;
+
+
+/**
+ * @brief The host address string extracted from the pre-signed URL.
+ *
+ * @note S3_PRESIGNED_GET_URL_LENGTH is set as the array length here as the
+ * length of the host name string cannot exceed this value.
+ */
+static char serverHost[ 256 ];
+
+/**
+ * @brief A buffer used in the demo for storing HTTP request headers and
+ * HTTP response headers and body.
+ *
+ * @note This demo shows how the same buffer can be re-used for storing the HTTP
+ * response after the HTTP request is sent out. However, the user can also
+ * decide to use separate buffers for storing the HTTP request and response.
+ */
+static uint8_t userBuffer[ USER_BUFFER_LENGTH ];
+
+/* The transport layer interface used by the HTTP Client library. */
+TransportInterface_t transportInterfaceHttp;
+
+/**
+ * @brief MQTT connection context used in this demo.
+ */
+static MQTTContext_t mqttContext;
+
+/**
+ * @brief Keep a flag for indicating if the MQTT connection is alive.
+ */
+static bool mqttSessionEstablished = false;
+
+/**
+ * @brief Structure for openssl parameters.
+ */
+static OpensslParams_t opensslParams;
+
+/**
+ * @brief Structure for openssl parameters.
+ */
+static OpensslParams_t opensslParamsHttp;
```

Comment:
```suggestion
/**
 * @brief Structure for openssl parameters for TLS session used by MQTT connection.
 */
static OpensslParams_t opensslParamsForMqtt;

/**
 * @brief Structure for openssl parameters for TLS session used by HTTP connection.
 */
static OpensslParams_t opensslParamsForHttp;
```

Reply:
Addressed in https://github.com/aws/aws-iot-device-sdk-embedded-C/pull/1425/commits/9f18e9905580d6e4067eaa4a215da9dbe1e2453a